examples/step-95/step-95.cc

```diff
@@ -0,0 +1,1711 @@
+    }
+
+    // With this function we can assemble the diagonal of the operator instead
+    // of applying the matrix-vector product. It uses the same function than the
+    // operator evalaution for the local operation on cells and faces but sets
+    // the template argument assemble to true.
+    void compute_diagonal(VectorType &diagonal) const
+    {
+      matrix_free->loop(&PoissonOperator::local_apply_cell<true>,
+                        &PoissonOperator::local_apply_face<true>,
+                        &PoissonOperator::local_apply_boundary_face,
+                        this,
+                        diagonal,
+                        diagonal);
+    }
+
+  private:
+    // We define two helper functions needed for matrix-free assembly
+    // operations.
+    template <typename Integrator>
+    void create_zero_basis(Integrator &evaluator) const
+    {
+      for (unsigned int i = 0; i < evaluator.dofs_per_cell; ++i)
+        evaluator.begin_dof_values()[i] = VectorizedArrayType(0.);
+    }
+
+    template <typename Integrator>
+    void create_standard_basis(const unsigned int j,
+                               Integrator        &evaluator) const
+    {
+      create_zero_basis(evaluator);
+      evaluator.begin_dof_values()[j] = VectorizedArrayType(1.);
+    }
+
+    // This function implements the local cell operation.
+    template <bool assemble = false>
+    void local_apply_cell(
+      const MatrixFree<dim, Number, VectorizedArrayType> &,
+      VectorType                                  &dst,
+      const VectorType                            &src,
+      const std::pair<unsigned int, unsigned int> &cell_range) const
+    {
+      CellIntegrator evaluator(*matrix_free, dof_index, quad_index);
+
+      const auto dofs_per_cell = evaluator.dofs_per_cell;
+
+      AlignedVector<VectorizedArrayType> dof_buffer(assemble ? dofs_per_cell :
+                                                               0);
+
+      const auto cell_range_category =
+        matrix_free->get_cell_range_category(cell_range);
+
+      // We define the cell_operation for inside cells as the standard Poisson
+      // term.
+      auto inside_cell_operation = [&](CellIntegrator &evaluator) {
+        evaluator.evaluate(EvaluationFlags::gradients);
+        for (unsigned int q : evaluator.quadrature_point_indices())
+          do_poisson_cell_term(evaluator, q);
+        evaluator.integrate(EvaluationFlags::gradients);
+      };
+
+      // We define the cell_operation for intersected cells as the standard
+      // Poisson term in the cut volume and the Nitsche term for weak Dirichlet
+      // boundary enforcement on the cut surface.
+      auto intersected_cell_operation = [&](CellIntegrator &evaluator) {
+        const unsigned int cell_batch_index =
+          evaluator.get_cell_or_face_batch_id();
+        const VectorizedArrayType diameter = cell_diameter[cell_batch_index];
+
+        const auto tau = compute_interior_penalty_parameter(diameter);
```

Comment:
Can you explain the choice of `tau`? If I read the code correctly, you set it in line 931 below to `5/diameter`. But I would have expected this to scale with the polynomial degree, as in regular IP methods for DG, where an scaling of size `degree * (degree + 1)` or `(degree + 1) * (degree + 1)` is common.

Reply:
I started to implement this for linear shape functions and apparently never updated it. Thank you for the hint, this now scales with the degree as it should.